### PR TITLE
Checking all dependencies are installed before starting threebot

### PR DESCRIPTION
**Pull request by AhmedSa-mir**
See https://github.com/threefoldtech/js-sdk/issues/64

### CODE REVIEW

File: jumpscale/servers/threebot/threebot.py [hunk 1]
```diff
@@ -468,6 +475,26 @@ def packages(self):
             self._packages = self._package_manager.get(self.instance_name)
         return self._packages
 
+    def check_dependencies(self):
+        install_msg = "Visit https://github.com/threefoldtech/js-sdk/blob/development/docs/wiki/quick_start.md for installation guide"
+
+        if not self.nginx.installed:
+            raise j.exceptions.NotFound(f"nginx is not installed.\n{install_msg}")
+        rc, _, _ = j.sals.process.execute("which certbot")
```
Comment (grimpy) [inline jumpscale/servers/threebot/threebot.py:483]: use shutil.which

File: jumpscale/servers/threebot/threebot.py [hunk 2]
```diff
@@ -468,6 +475,26 @@ def packages(self):
...
+        if rc:
+            raise j.exceptions.NotFound(f"certbot is not installed.\n{install_msg}")
+        rc, _, _ = j.sals.process.execute("dpkg -l python-certbot-nginx")
```
Comment (grimpy) [inline jumpscale/servers/threebot/threebot.py:486]: if it's python you should be able to check by attempting an import or via resource_pkg instead

File: jumpscale/tools/redis/redis.py [hunk 3]
```diff
@@ -25,8 +25,10 @@ def start(self, host: str = "127.0.0.1", port: int = 6379):
             host (str, optional): redis bind address. Defaults to "127.0.0.1".
             port (int, optional): redis port. Defaults to 6379.
         """
-        self.cmd.start_cmd = f"redis-server --bind {host} --port {port}"
-        self.cmd.start()
+        # Port is not busy (Redis is not started)
+        if not j.sals.nettools.tcp_connection_test("127.0.0.1", 6379, timeout=1):
+            self.cmd.start_cmd = f"redis-server --bind {host} --port {port}"
+            self.cmd.start()
```
Comment (waleedhammam) [inline jumpscale/tools/redis/redis.py:31]: should add port/process string/reges to be able to hold the process for later stop

Comment (waleedhammam) [reply jumpscale/tools/redis/redis.py:31]: check nginx or threebot entry point

File: jumpscale/servers/threebot/threebot.py (hunk 2 above)
Comment (AhmedSa-mir) [reply jumpscale/servers/threebot/threebot.py:487]: It's an nginx plugin and installed using apt

Comment (grimpy) [reply jumpscale/servers/threebot/threebot.py:487]: it's not an nginx plugin, it's infact a plugin for certbot to be able to manage nginx.
You can check if this plugin is installed by doing
try:
   import certbot_nginx
   its installed
except ImportError:
  no it's not installed

File: jumpscale/tools/nginx/nginxserver.py [hunk 4]
```diff
@@ -13,21 +13,14 @@ def __init__(self, *args, **kwargs):
...
-            "apt-get install -y software-properties-common;"
-            "add-apt-repository universe;"
-            "add-apt-repository ppa:certbot/certbot;"
-            "apt-get update;"
-            "apt-get install -y nginx certbot python-certbot-nginx;",
-            showout=True,
-            die=False,
-            env={"DEBIAN_FRONTEND": "noninteractive"},
-        )
+        return j.sals.process.execute("which nginx")[0] == 0
```
Comment (grimpy) [inline jumpscale/tools/nginx/nginxserver.py:23]: also here shutil.which

File: jumpscale/tools/redis/redis.py [hunk 5]
```diff
@@ -16,7 +16,7 @@ def installed(self) -> bool:
         Returns:
             bool: True if redis server is installed
         """
-        return j.sals.process.execute('which redis-server')[0] == 0
+        return j.sals.process.execute("which redis-server")[0] == 0
```
Comment (grimpy) [inline jumpscale/tools/redis/redis.py:19]: shutil.which

**Pull request merged**